### PR TITLE
added `view_as`

### LINKED ISSUE #328: Impement TensorBase.view_as
### 🚀 Feature

Implement [`TensorBase.view_as`](https://pytorch.org/docs/stable/generated/torch.Tensor.view_as.html)

### Motivation

NeMo [Vision Transformer](https://github.com/NVIDIA/NeMo/blob/main/examples/vision/vision_transformer/megatron_vit_classification_pretrain.py)

cc @tfogal

### PR DESCRIPTION
**Pull request by k223kim**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #328. I am not sure if I should include the test case for `view_as` `opinfos.py` (cause there wasn't one for `view`, maybe it is similar to `reshape`?). Also, I added `view_as` to prims, although I am not sure if this is necessary. Let me know how I can improve!

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

👊

### CODE REVIEW

File: thunder/core/prims.py [hunk 1]
```diff
@@ -3187,6 +3188,7 @@ def transpose_meta(a: TensorProxy, /, permutation: tuple[int, ...]) -> TensorPro
 
 
 view = make_prim(PrimIDs.VIEW, "view", meta=reshape_meta, tags=(OpTags.SHAPE_OP,))
+view_as = make_prim(PrimIDs.VIEW_AS, "view_as", meta=reshape_meta, tags=(OpTags.SHAPE_OP,))
```
Comment (mruberry) [inline thunder/core/prims.py:3191]: Let's not add a new primitive -- it won't be called by the implementation, anyway

File: thunder/executors/torchex.py [hunk 2]
```diff
@@ -539,6 +540,7 @@ def _squeeze_transform(a: TensorLike, /, dim: None | int | Sequence[int] = None)
 _register_implementation(prims.transpose, checker=_always_executable, execution_transform=_transpose_prim_transform)
 _register_implementation(prims.unfold, unfold, checker=_always_executable)
 _register_implementation(prims.view, view, checker=_always_executable)
+_register_implementation(prims.view_as, view_as, checker=_always_executable)
```
Comment (mruberry) [inline thunder/executors/torchex.py:543]: Without the primitive, we can remove this registration

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -1031,6 +1031,11 @@ def view(a: TensorLike, /, *shape) -> TensorLike:
     return reshape(a, shape)
 
 
+@torchsymbol(torch.Tensor.view_as, is_method=True)
+def view_as(a: TensorLike, /, b: TensorLike) -> TensorLike:
```
Comment (mruberry) [inline thunder/torch/__init__.py:1035]: nit: `a: TensorLike, b: TensorLike, /)` — the reason for this is protect us from people writing code like `view_as(b=...)` and having to support the parameter name "b" forever, even though the name doesn't really matter in this case

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -1031,6 +1031,11 @@ def view(a: TensorLike, /, *shape) -> TensorLike:
...
+    return view(a, b.size())
```
Comment (mruberry) [inline thunder/torch/__init__.py:1036]: Nice implementation!

**Review by mruberry — APPROVED** [review]
Hey @k223kim! Cool stuff! I made a couple minor comments

File: thunder/core/prims.py (hunk 1 above)
Comment (k223kim) [reply thunder/core/prims.py:3191]: Just for my understanding, when do we have to add a new primitive? Is there a general rule of thumb?

File: thunder/torch/__init__.py (hunk 3 above)
Comment (k223kim) [reply thunder/torch/__init__.py:1035]: Oh my bad. I'll make sure to check on this next time! Thanks!

**Comment by k223kim** [comment]
Hi @mruberry! Thanks for the awesome review! Just some questions regarding this PR. I have added a test case for `view_as` but I am not sure if I should include `supports_grad=True`. If I do so, it throws the following errors:
```
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_view_as_torch_cpu_bfloat16 - RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.
```
Do you think this is necessary?

**Comment by mruberry** [comment]
> Hi @mruberry! Thanks for the awesome review! Just some questions regarding this PR. I have added a test case for `view_as` but I am not sure if I should include `supports_grad=True`. If I do so, it throws the following errors:
> 
> ```
> FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_view_as_torch_cpu_bfloat16 - RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.
> ```
> 
> Do you think this is necessary?

Nope, don't need to add `supports_grad=True` for now. fyi @IvanYashchuk

**Review by mruberry — APPROVED** [review]
Cool! Looks great with thorough tests

fyi @Borda and @t-vi to merge

**Comment by t-vi** [comment]
Thank you @k223kim @mruberry

**Pull request merged**